File: apps/api/tests/_local_supabase.py
# Shared helpers for tests that exercise a real local Supabase stack
# (`supabase start`), per STANDARDS.md: "No mocking of Supabase in
# integration tests — use `supabase start` for a local instance."
#
# These are the Supabase CLI's fixed local-dev demo keys — provisioned by
# every `supabase start`, not secret (sourced here from the running local
# containers' own kong.yml routing config, not guessed). They must never
# be the keys used against the live project (those live in `.env` and are
# real secrets) — conftest.py overrides SUPABASE_URL/SUPABASE_SERVICE_ROLE_KEY
# to these local values before `main.app` is ever imported, so the whole
# test suite talks to the local stack, never the live one.

import uuid

import httpx
from supabase import Client, create_client

LOCAL_SUPABASE_URL = "http://127.0.0.1:54321"
LOCAL_SUPABASE_ANON_KEY = (
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9."
    "eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9."
    "CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0"
)
LOCAL_SUPABASE_SERVICE_ROLE_KEY = (
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9."
    "eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0."
    "EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU"
)
LOCAL_POSTGRES_DSN = "postgresql://postgres:postgres@localhost:54322/postgres"


def admin_client() -> Client:
    """Service-role client — bypasses RLS, same client the app itself
    uses (db.client.get_service_role_client points at the same local
    values once conftest.py's env override is in effect)."""
    return create_client(LOCAL_SUPABASE_URL, LOCAL_SUPABASE_SERVICE_ROLE_KEY)


def create_test_user(client: Client, *, password: str = "test-password-123") -> tuple[str, str]:
    """Creates a real, auto-confirmed local auth user. Returns (user_id, email)."""
    email = f"test-{uuid.uuid4().hex[:12]}@example.com"
    result = client.auth.admin.create_user({"email": email, "password": password, "email_confirm": True})
    return result.user.id, email


def delete_test_user(client: Client, user_id: str) -> None:
    """Cascades to that user's documents/chunks via the schema's `on
    delete cascade` FKs — does not touch storage objects, which have no
    such FK; callers that upload files must clean those up separately."""
    client.auth.admin.delete_user(user_id)


def login(email: str, password: str = "test-password-123") -> str:
    """Real password-grant login against local GoTrue — returns a
    genuinely issued ES256 access token, not a self-forged one (see
    .agent/MEMORY.md's anti-pattern entry on circular JWT verification)."""
    response = httpx.post(
        f"{LOCAL_SUPABASE_URL}/auth/v1/token?grant_type=password",
        json={"email": email, "password": password},
        headers={"apikey": LOCAL_SUPABASE_ANON_KEY, "Content-Type": "application/json"},
        timeout=10,
    )
    response.raise_for_status()
    return response.json()["access_token"]


def upload_via_rest(access_token: str, bucket: str, path: str, content: bytes, content_type: str) -> httpx.Response:
    """Uploads as the authenticated user over the real Storage REST API —
    exercises the same storage.objects RLS policies (and HTTP path) a
    real browser client would hit, not a service-role bypass."""
    return httpx.put(
        f"{LOCAL_SUPABASE_URL}/storage/v1/object/{bucket}/{path}",
        content=content,
        headers={
            "apikey": LOCAL_SUPABASE_ANON_KEY,
            "Authorization": f"Bearer {access_token}",
            "Content-Type": content_type,
        },
        timeout=30,
    )


def rest_select(access_token: str, table: str, query: str = "") -> httpx.Response:
    """RLS-scoped PostgREST select as the authenticated user — used to
    prove multi-tenant isolation at the same layer the app itself reads
    through, independent of any /documents endpoint (FEAT-008's job, not
    built yet)."""
    url = f"{LOCAL_SUPABASE_URL}/rest/v1/{table}"
    if query:
        url = f"{url}?{query}"
    return httpx.get(
        url,
        headers={"apikey": LOCAL_SUPABASE_ANON_KEY, "Authorization": f"Bearer {access_token}"},
        timeout=10,
    )

File: apps/web/e2e/_local-supabase.ts
// Mirrors apps/api/tests/_local_supabase.py — these are the Supabase CLI's
// fixed local-dev demo keys (provisioned by every `supabase start`), not
// secret. Used only to clean up test users created against the local stack
// during these e2e runs; must never point at the live project.

export const LOCAL_SUPABASE_URL = "http://127.0.0.1:54321";
export const LOCAL_SUPABASE_SERVICE_ROLE_KEY =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9." +
  "eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0." +
  "EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU";

export async function deleteTestUserByEmail(email: string): Promise<void> {
  const listRes = await fetch(
    `${LOCAL_SUPABASE_URL}/auth/v1/admin/users?email=${encodeURIComponent(email)}`,
    {
      headers: {
        apikey: LOCAL_SUPABASE_SERVICE_ROLE_KEY,
        Authorization: `Bearer ${LOCAL_SUPABASE_SERVICE_ROLE_KEY}`,
      },
    }
  );
  if (!listRes.ok) return;
  const { users } = (await listRes.json()) as { users: { id: string }[] };
  for (const user of users) {
    await fetch(`${LOCAL_SUPABASE_URL}/auth/v1/admin/users/${user.id}`, {
      method: "DELETE",
      headers: {
        apikey: LOCAL_SUPABASE_SERVICE_ROLE_KEY,
        Authorization: `Bearer ${LOCAL_SUPABASE_SERVICE_ROLE_KEY}`,
      },
    });
  }
}

export async function createTestUser(
  email: string,
  password: string
): Promise<string> {
  const res = await fetch(`${LOCAL_SUPABASE_URL}/auth/v1/admin/users`, {
    method: "POST",
    headers: {
      apikey: LOCAL_SUPABASE_SERVICE_ROLE_KEY,
      Authorization: `Bearer ${LOCAL_SUPABASE_SERVICE_ROLE_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email, password, email_confirm: true }),
  });
  if (!res.ok) {
    throw new Error(`createTestUser failed: ${res.status} ${await res.text()}`);
  }
  const { id } = (await res.json()) as { id: string };
  return id;
}

// The local stack's mail catcher is Mailpit (confirmed live during the
// FEAT-013 auth self-audit) — the `INBUCKET_URL`/`MAILPIT_URL` printed by
// `supabase status` point at the same host:port; only Mailpit's own REST API
// (/api/v1/messages) actually works against it, not classic Inbucket's.
export const LOCAL_MAILPIT_URL = "http://127.0.0.1:54324";

interface MailpitMessageSummary {
  ID: string;
  To: { Address: string }[];
}

interface MailpitMessagesResponse {
  messages: MailpitMessageSummary[];
}

interface MailpitMessage {
  Text: string;
}

/** Polls Mailpit for the most recent message to `email` and returns the
 * first http(s) link found in its plain-text body — used to pick up real
 * password-reset/confirmation links a real GoTrue send actually produced. */
export async function waitForEmailLink(
  email: string,
  { timeoutMs = 8000, intervalMs = 300 }: { timeoutMs?: number; intervalMs?: number } = {}
): Promise<string | null> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const listRes = await fetch(`${LOCAL_MAILPIT_URL}/api/v1/messages`);
    if (listRes.ok) {
      const data = (await listRes.json()) as MailpitMessagesResponse;
      const match = data.messages.find((m) => m.To.some((t) => t.Address === email));
      if (match) {
        const msgRes = await fetch(`${LOCAL_MAILPIT_URL}/api/v1/message/${match.ID}`);
        if (msgRes.ok) {
          const full = (await msgRes.json()) as MailpitMessage;
          const linkMatch = full.Text.match(/https?:\/\/[^\s)]+/);
          if (linkMatch) return linkMatch[0];
        }
      }
    }
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
  return null;
}
